flang/lib/Lower/Allocatable.cpp

```diff
@@ -221,7 +248,7 @@ class AllocateStmtHelper {
       genDescriptorRollBack(boxAddr, backupBox);
   }
 
-  mlir::Value genDescriptorBackup(mlir::Value boxAddr) {
+  mlir::Value genDescriptorBackup(fir::BoxAddressValue boxAddr) {
     // back-up descriptors in case something goes wrong. This is to fullfill
     // Fortran 2018 9.7.4 point 6 requirements that the original descriptor is
```

Comment:
? Bullet 6 is about STAT variables used in ALLOCATE/DEALLOCATE statements.

In the event of an error other than FAILED IMAGE, I read this clause to mean that there are exactly two outcomes, all relating to the storage pointer buried in the descriptor.

1. The allocation (deallocation) is successful and the pointer (allocatable) base reference is updated. 
2. The allocation (deallocation) fails and the pointer (allocatable) base reference is not updated.

It doesn't really say the entire descriptor update should be treated as a transaction in memory.

Reply:
I think you are right here, I over-interpreted what is the _allocation status_ / _pointer association status_ in 
> each allocate-object that was not successfully allocated or deallocated shall retain its previous allocation status or pointer association status.

My understanding was that it meant that all the allocatable/pointer properties (e.g. bounds, length parameters, base address, and even the type for polymorphic allocatables) should be left untouched in case allocation/deallocation failed.
Because that's non trivial to keep track of all the changes that are made to the descriptor until the last moment an allocation can fail (which for derived types can happen quite late with derived type allocatable with initialized components that themselves require allocation that can fail), I wanted to back-up the descriptor.
But reading  at what is _allocation status_ again, I think you are right and there is not real need to do this. It seems ok if the allocatable/pointer is in a zombie state (wrong bounds, type...) as long as we can answer allocated/not allcoated correctly.

Thanks, that will simplify error recovery quite a bit.